tower-http/src/lib.rs

```diff
@@ -1,3 +1,83 @@
+//! `async fn(HttpRequest) -> Result<HttpResponse, Error>`
+//!
+//! # Overview
+//!
+//! tower-http is a library that provides HTTP specific middlewares and utilities built on top of
+//! [tower].
+//!
+//! All middlewares uses the [http] and [http-body] crates as the HTTP abstractions. That means
+//! they're compatible with any library or framework that also uses those crates, such as [hyper].
+//! Some middlewares might bring in other dependencies for doing various things.
+//!
+//! # Example
+//!
+//! ```rust
+//! use tower_http::{
+//!     add_extension::AddExtensionLayer,
+//!     compression::CompressionLayer,
+//! };
+//! use tower::{ServiceBuilder, service_fn};
+//! use http::{Request, Response};
+//! use hyper::Body;
+//! use std::sync::Arc;
+//! # struct Error;
+//! # struct DatabaseConnectionPool;
+//! # impl DatabaseConnectionPool {
+//! #     fn new() -> DatabaseConnectionPool { DatabaseConnectionPool }
+//! # }
+//! # async fn run_http_service<T>(_: T) {}
+//!
+//! async fn handler(request: Request<Body>) -> Result<Response<Body>, Error> {
+//!     // ...
+//!     # todo!()
+//! }
+//!
+//! struct State {
+//!     pool: DatabaseConnectionPool,
+//! }
+//!
+//! #[tokio::main]
+//! async fn main() {
+//!     let state = State {
+//!         pool: DatabaseConnectionPool::new(),
+//!     };
+//!
+//!     let service = ServiceBuilder::new()
+//!         // Share an `Arc<State>` with all requests
+//!         .layer(AddExtensionLayer::new(Arc::new(state)))
+//!         // Compress responses
+//!         .layer(CompressionLayer::new())
+//!         // Wrap a `Service` in our middleware stack
+//!         .service(service_fn(handler));
+//!
+//!     // Run our service using some HTTP server
+//!     run_http_service(service).await;
```

Comment:
i'd kind of like to see a few more comments explaining what this code is doing. for example:
```suggestion
//!
//! // Our request handler. This is where we would implement the application logic
//! // for responding to HTTP requests...
//! async fn handler(request: Request<Body>) -> Result<Response<Body>, Error> {
//!     // ...
//!     # todo!()
//! }
//!
//! /// Shared state across all request handlers --- in this case, a pool of database connections.
//! struct State {
//!     pool: DatabaseConnectionPool,
//! }
//!
//! #[tokio::main]
//! async fn main() {
//!     // Construct the shared state.
//!     let state = State {
//!         pool: DatabaseConnectionPool::new(),
//!     };
//!
//!     // Use `tower`'s `ServiceBuilder` API to build a stack of `tower` middleware
//!     // wrapping our request handler.
//!     let service = ServiceBuilder::new()
//!         // Share an `Arc<State>` with all requests
//!         .layer(AddExtensionLayer::new(Arc::new(state)))
//!         // Compress responses
//!         .layer(CompressionLayer::new())
//!         // Wrap a `Service` in our middleware stack
//!         .service(service_fn(handler));
//!
//!     // Run our service using some HTTP server.
//!     // The HTTP server implementation is provided by a compatible external library,
//!     // such as `hyper`.
//!     run_http_service(service).await;
```